proto/cosmos/upgrade/v1beta1/upgrade.proto

```diff
@@ -40,6 +40,39 @@ message Plan {
   // If this field is not empty, an error will be thrown.
   google.protobuf.Any upgraded_client_state = 5
       [deprecated = true];
+
+  // Optional: Upgrade contains additional instructions for the devops or a hypervisor.
+  // App specific instructions are handled by the `info` attribute. Here we provide
+  // information such as pre-upgrade or post-upgrade commands.
+  UpgradeInstructions Upgrade = 6;
+}
+
+message UpgradeInstructions {
+  option (gogoproto.equal)            = true;
+
+  // If not empty, a shell command to be run by the upgrade manager or a hypervisor after shutting down
+  // the app and before running a new node.
+  string pre_run = 1;
+  // If not empty, a shell command to be run by the upgrade manager or a hypervisor after shutting down
+  // the app and after running a new app.
+  string post_run = 2;
+  // List of required assets to download. This follows the cosmovisor structure.
+  // SHOULD have only one entry per platform.
+  repeated Asset assets    = 3;
+  // Description contains additional information about the upgrade process. Can reference an external resource.
+  string description = 4;
+}
+
+message Asset {
+  option (gogoproto.equal)            = true;
+
+  // Platform identifier. It's composed from OS and CPU architecture. Example: "linux/amd64"
+  string platform = 1;
+  // URL to a script or binary to download. Should be related to the UpgradeInstructions pre_run or post_run
+  // commands. If multiple files are needed, then they should be packed in a gzip archive.
+  string url = 2;
+  // Checksum is a sha256 hex encoded checksum of an artifact referenced by the URL.
+  string checksum = 3;
 }
```

Comment:
With the addition of the `pre_run` and `post_run` fields, there will be a need to download those as well. By being able to provide a name for assets, users won't be constrained to only providing archives when they need to include a special `pre_run` script.

I feel like adding a `string name` field to the `Asset` is the most versatile solution. We'd state that entries in `repeated Asset assets` SHOULD be unique by `name` and `platform` (as opposed to just `platform`).

Alternatively, a hierarchy could be used, but there's a couple ways to do that, either name -> platforms or platform -> names. There's good arguments for either way though. But I feel like leaving it flat allows users to handle it the way that makes the most sense for them.

### Cosmovisor Usage

Assets would be downloaded in the order provided but only if there is a `Source` entry that matches the system's `platform` (or there's a `platform` = `any` entry for it). Basically, we'd loop through the `assets`, if the `platform` matches the system, then it's downloaded. Or if the `platform` is `any` and the `name` is one that hasn't yet been downloaded this upgrade, then it'll be downloaded.

If the `url` returns an archive, then the `name` becomes a directory name. Otherwise it's a filename. Either way, the full path to the downloaded result would be `{DAEMON_HOME}/{upgrade name}/{name}`. Sub directories would be allowed in the `name`, but we'd need to discuss whether or not to allow it to contain `../`.

I'd make a couple special cases for the `name` value:
1. If the `name` is empty, and the `url` returns an archive, it is unpacked and becomes `{DAEMON_HOME}/{upgrade name}` (similar to current behavior). If the `name` is empty and the `url` returns a non-archive file, it'll end up as `{DAEMON_HOME}/{upgrade name}/bin/{DAEMON_NAME}`.
2. If the `name` is `.` then the `url` will need to return an archive and it'll be unpacked into `{DAEMON_HOME}/{upgrade name}`. If the name is `.` and the `url` returns something other than an archive, an error will be given and the upgrade/download process will be interrupted.

We'd probably want to maintain the current functionality where, after doing the downloads, if `{DAEMON_HOME}/{upgrade name}/bin/{DAEMON_NAME}` doesn't exist, but `{DAEMON_HOME}/{upgrade name}/{DAEMON_NAME}` does, the latter is copied to the former.